core/logger/zap.go

```diff
@@ -0,0 +1,114 @@
+package logger
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+var _ Logger = &zapLogger{}
+
+type zapLogger struct {
+	*zap.SugaredLogger
+	config zap.Config
+	name   string
+	fields []interface{}
+}
+
+func newZapLogger(cfg zap.Config) (Logger, error) {
+	zl, err := cfg.Build()
+	if err != nil {
+		return nil, err
+	}
+	return &zapLogger{config: cfg, SugaredLogger: zl.Sugar()}, nil
+}
+
+func (l *zapLogger) SetLogLevel(lvl zapcore.Level) {
+	l.config.Level.SetLevel(lvl)
+}
+
+func (l *zapLogger) With(args ...interface{}) Logger {
+	newLogger := *l
+	newLogger.SugaredLogger = l.SugaredLogger.With(args...)
+	newLogger.fields = copyFields(l.fields, args...)
+	return &newLogger
+}
+
+// copyFields returns a copy of fields with add appended.
+func copyFields(fields []interface{}, add ...interface{}) []interface{} {
+	f := make([]interface{}, 0, len(fields)+len(add))
+	f = append(f, fields...)
+	f = append(f, add...)
+	return f
+}
+
+func joinName(old, new string) string {
+	if old == "" {
+		return new
+	}
+	return old + "." + new
+}
+
+func (l *zapLogger) Named(name string) Logger {
+	newLogger := *l
+	newLogger.name = joinName(l.name, name)
+	newLogger.SugaredLogger = l.SugaredLogger.Named(name)
+	newLogger.Trace("Named logger created")
```

Comment:
I had already been thinking about adding a way to dump a map of these injected Named loggers anyways since it's a useful map of the system (like how some http router packages can print all the routes to generate docs).